message_test.go

```diff
@@ -19,7 +20,11 @@ func TestMessage(t *testing.T) {
 		Labels:    map[string]string{"source": "test"},
 		Message:   "Hello world!",
 	}
-	if m.String() != "2006-01-02T15:04:05Z\terror\tsource=test\tHello world!" {
-		t.Fatalf("Incorrect message string: %s", m.String())
-	}
+	assert.Equals(
```

Comment:
I'm just surprised that this is running in an environment where arca thinks that `stderr` is a *terminal*. `go test` apparently redirects the CUT `stderr` to the `go test` `stdout`, but that implies that a GitHub Action's `stdout` behaves like a `tty` ... which seems surprising. (A few searches didn't turn up anything useful.)

The thing is ... I'm not sure whether this is a bizarre curiosity, or whether it suggests a problem in your use (or Go's implementation!) of `isTerminal`, and that worries me. 😟 A bit...

Reply:
I've lost my grip:  the code here (er, below this point) doesn't care whether the output device is a terminal(-like) or not -- it's just testing whether the output matches what the logger produces, which may or may not have the escape sequences in it, depending on what the values of `log.SetColor` and `log.ResetColor` happen to end up being.

Now, if there were assertions testing _those_ values, that would be interesting...but, I'm not exactly sure that we care...maybe...um...hmmm....

Reply:
What I'd like to see, ideally, is validation that log output to a non-terminal doesn't include the escape sequences while output to a terminal does include them.

There's also "verify that the log message itself is right regardless of the ANSI wrapping paper", which this does nicely; but it's not really a test of the new code.

Reply:
What I ran into previously with this is that a local `go test` seemed to act like a terminal, but the github action acted like not-a-terminal. So yeah, this change works around (or though?) that difference by including the `log.SetColor` and `log.ResetColor` values, which passes both locally and in the action. But of course, to Dave's point, this doesn't really test the new code. If there is a way for me to force the test into the terminal or non-terminal conditions, I'm happy to add that, but I'm not sure what that would look like and whether it might include another external dependency that is frowned upon.

Reply:
One approach is to provide an external control that the test harness (and perhaps the `arcaflow` command) can use to force the choice (i.e., an override of the `term.IsTerminal()` result).  (That could be a global variable, or a functional setter, or an environment variable, etc., but it would require more code and infrastructure.)

The other approach is to control the environment that the test is executed in.  For example, have the test create a subprocess to run the scenario in, where it can control the identity of `stderr`.  (I don't know if a pipe counts as a terminal -- I would presume that it doesn't -- but, if necessary, `stderr` could be redirected to a disk file and read back in.)

None of this is hard, but, as Matt is fond of saying, I'm not sure the juice is worth the squeeze.

> a local `go test` seemed to act like a terminal, but the github action acted like not-a-terminal.

I'm guessing that you were executing `go test` interactively, and therefore `stderr` was directed to your terminal; whereas, the GitHub Action is run non-interactively, without a terminal, with `stderr` directed to a pipeline ultimately ending up in a disk file or similar...with no terminal (or pseudo-tty) in the line.  No surprises here.  Perhaps you would get a consistent result if you redirected output from your interactive command to a file, e.g.:
```
go test >test.out 2>&1
```

Reply:
Ah: that makes sense. Locally I'd expect `go test`'s `stderr` to be your `/dev/tty`, whereas I think I'm kinda glad to hear that in the GitHub action it's a file.

One way to finesse this would be to make the test code *independently* based on `term.isTerminal`, so that it expects color for a terminal (interactively) and no color for the CI (non-interactively). For right now, assuming you verify it interactively, we'll know it's been tested both ways ... in the longer run, we'd expect it to be normally run only under the CI in dull monochrome, and that's a bit unsatisfying.

The problem is that would require firing up `go test` under a tty within the CI and scraping the output, maybe with `screen`. That would be the only way we can be sure the code works correctly in both modes.

Reply:
> One way to finesse this would be to make the test code independently based on `term.isTerminal`, so that it expects color for a terminal (interactively) and no color for the CI (non-interactively).

The thing is, that solves only half the problem, and it pretty much guarantees that the interactive experience will be different from the CI one, which is always great fun.

As you say, the "right" way to test this would be for the test harness to run the test as a subprocess with and without a pty (and `screen` might be the most straightforward way to provide a pty).

However, I'd be inclined to make the test work correctly for the CI/non-interactive case, assuming that the developers and users will have enough exposure to the tty case that testing it is unimportant, but that would likely break the developer trying to run the tests interactively.  So, maybe your idea of having the test determine for itself what it should see at runtime is the most expedient.

Reply:
Really, since this is a unit test, working within the `log` package, the best solution would be Webb's earlier comment about making it a true unit test and mocking the external `term.isTerminal` to operate both ways regardless of the *actual* `stderr`. I have no idea how to do that in the Go test environment.